ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp

```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <vector>
+
+#include "moreh_clip_grad_norm_step2_device_operation.hpp"
+#include "tt_metal/common/work_split.hpp"
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+#include "ttnn/operations/moreh/moreh_helper_functions.hpp"
+
+namespace ttnn::operations::moreh::moreh_clip_grad_norm_step2 {
+
+std::tuple<uint32_t, float, bool> get_p_decimal_p_is_negative(float ord) {
+    auto p = std::floor(ord);
+    auto decimal = ord - p;
+    const bool p_is_negative = p < 0.0f;
+    if (p_is_negative) {
+        p = -p;
+    }
+    return std::make_tuple(static_cast<uint32_t>(p), decimal, p_is_negative);
+}
+
+MorehClipGradNormStep2Operation::ProgramFactory::cached_program_t
+MorehClipGradNormStep2Operation::ProgramFactory::create(
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& total_norm) {
+    const auto& tmp_pow_sum = tensor_args.tmp_pow_sum;
+    auto norm_type = operation_attributes.norm_type;
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                      Device Setup
+    ////////////////////////////////////////////////////////////////////////////
+    auto device = tmp_pow_sum.device();
+    auto program = CreateProgram();
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                         Parameters Setup
+    ////////////////////////////////////////////////////////////////////////////
+    const auto num_tiles = tmp_pow_sum.volume() / tt::constants::TILE_HW;
+
+    auto [p, decimal, p_is_negative] = get_p_decimal_p_is_negative(1.0f / norm_type);
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                         Core Setup
+    ////////////////////////////////////////////////////////////////////////////
+    CoreCoord single_core = {0, 0};
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                         CircularBuffer Setup
+    ////////////////////////////////////////////////////////////////////////////
+    const uint32_t in0_t = 1;  // input(==tmp_pow_sum)
+    const uint32_t in1_t = 1;  // decimal
+
+    // x^p * exp(log(x) * decimal)
+    const uint32_t out0_t = 1;  // output(==total_norm)
+
+    const uint32_t im0_t = 1;  // Sum[tmp_pow_sum](==x)
+    const uint32_t im1_t = 1;  // x^p
+    const uint32_t im2_t = 1;  // log(x)
+    const uint32_t im3_t = 1;  // exp(log(x) * decimal)
+
+    const auto cb_data_format = tt::tt_metal::datatype_to_dataformat_converter(total_norm.get_dtype());
+
+    ttnn::operations::CreateCircularBuffer(
```

Comment:
```suggestion
    CreateCircularBuffer(
```

Reply:
Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)